src/System.Threading/tests/AsyncLocalTests.cs

```diff
@@ -419,5 +472,180 @@ private static Task Run(Func<Task> func)
                 }
             }
         }
+
+        [Theory]
+        [MemberData(nameof(GetCounts))]
+        public static async Task AsyncLocalsUnwind(int count)
+        {
+            AsyncLocal<object>[] asyncLocals = new AsyncLocal<object>[count];
+
+            ExecutionContext Default = ExecutionContext.Capture();
+            int[] manuallySetCounts = new int[count];
+            int[] automaticallyUnsetCounts = new int[count];
+            int[] automaticallySetCounts = new int[count];
+            ExecutionContext[] capturedContexts = new ExecutionContext[count];
+
+            // Setup the AsyncLocals; capturing ExecutionContext for each level
+            await SetLocalsRecursivelyAsync(count - 1);
+
+            ValidateCounts(thresholdIndex: 0, maunalSets: 1, automaticUnsets: 1, automaticSets: 0);
+            ValidateAsyncLocalsValuesNull();
+
+            // Check Running with the contexts captured when setting the locals
+            TestCapturedExecutionContexts();
+
+            ExecutionContext.SuppressFlow();
+            try
+            {
+                // Re-check restoring, but starting with a suppressed flow
+                TestCapturedExecutionContexts();
```

Comment:
This caught me with a null ref in "Reduce Execution Context Save+Restore" fix https://github.com/dotnet/coreclr/pull/15629/commits/c11e2e1736e28c6d7b57165a065d2f109b193dc8

In a test that wasn't for async locals